### DISCUSSION TITLE
WNN analysis of 10x Multiome, RNA + ATAC

### DISCUSSION THREAD

**Discussion by drkoryjohns** [satijalab/seurat#6552] [Q&A]
Hello Seurat community!

Need guidance please.

I have 8 separate paired GEX and ATAC samples that I would like to analyze together.

To do, I combined the GEX data using harmony.

For merging the ATAC data, I followed the workflow here:

https://github.com/stuart-lab/signac/blob/master/vignettes/merging.Rmd

I was able to transfer cell type annotations from GEX to ATAC and create a co-embedded umap using the workflow here:

https://satijalab.org/seurat/articles/atacseq_integration_vignette.html#co-embedding-scrna-seq-and-scatac-seq-datasets-1

However, I am running into problem of trying to do apply the WNN workflow and downstream motif and differential analyses found here:

https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html#wnn-analysis-of-10x-multiome-rna-atac-1

Specifically, this workflow demonstrates how to create a combined Seurat object for one sample pair of GEX and ATAC data.

How do I combine multiple chrom.assays into one so that I can assign the combined chrom.assays to the [[ATAC]] assay slot of the Seurat object expected for the WNN workflow?  Or, how can I use the merged Seurat object to extract a single chrom.assay that contains all the data across the 8 samples?

Thank you!

My code is below if you should like to inspect:

#------------------------------------------------------------------------------
# ATAC Analysis
# https://github.com/stuart-lab/signac/blob/master/vignettes/merging.Rmd
#------------------------------------------------------------------------------

plan("multiprocess", workers = 4)
options(future.globals.maxSize = 50000 * 1024^2) # for 50 Gb RAM

DG_F1_VV_atac_bed <- read.table("/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/bed_files/DG_F1_VV_atac_peaks.bed",col.names = c("chr", "start", "end"))

DG_F2_VV_atac_bed <- read.table("/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/bed_files/DG_F2_VV_atac_peaks.bed",col.names = c("chr", "start", "end"))

DG_M1_VV_atac_bed <- read.table("/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/bed_files/DG_M1_VV_atac_peaks.bed",col.names = c("chr", "start", "end"))

DG_M2_VV_atac_bed <- read.table("/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/bed_files/DG_M2_VV_atac_peaks.bed",col.names = c("chr", "start", "end"))

NAC_F1_VV_atac_bed <- read.table("/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/bed_files/NAC_F1_VV_atac_peaks.bed",col.names = c("chr", "start", "end"))

NAC_M1_VV_atac_bed <- read.table("/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/bed_files/NAC_M1_VV_atac_peaks.bed",col.names = c("chr", "start", "end"))

VH1PP_F1_VV_atac_bed <- read.table("/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/bed_files/VH1PP_F1_VV_atac_peaks.bed",col.names = c("chr", "start", "end"))

VH1PP_M1_VV_atac_bed <- read.table("/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/bed_files/VH1PP_M1_VV_atac_peaks.bed",col.names = c("chr", "start", "end"))

dim(DG_F1_VV_atac_bed)
dim(DG_F2_VV_atac_bed)
dim(DG_M1_VV_atac_bed)
dim(DG_M2_VV_atac_bed)
dim(NAC_F1_VV_atac_bed)
dim(NAC_M1_VV_atac_bed)
dim(VH1PP_F1_VV_atac_bed)
dim(VH1PP_M1_VV_atac_bed)

gr.DG.F1.VV <- makeGRangesFromDataFrame(DG_F1_VV_atac_bed)
gr.DG.F2.VV <- makeGRangesFromDataFrame(DG_F2_VV_atac_bed)
gr.DG.M1.VV <- makeGRangesFromDataFrame(DG_M1_VV_atac_bed)
gr.DG.M2.VV <- makeGRangesFromDataFrame(DG_M2_VV_atac_bed)
gr.NAC.F1.VV <- makeGRangesFromDataFrame(NAC_F1_VV_atac_bed)
gr.NAC.M1.VV <- makeGRangesFromDataFrame(NAC_M1_VV_atac_bed)
gr.VH1PP.F1.VV <- makeGRangesFromDataFrame(VH1PP_F1_VV_atac_bed)
gr.VH1PP.M1.VV <- makeGRangesFromDataFrame(VH1PP_M1_VV_atac_bed)

combined.peaks <- reduce(x=c(gr.DG.F1.VV,
gr.DG.F2.VV,
gr.DG.M1.VV,
gr.DG.M2.VV,
gr.NAC.F1.VV,
gr.NAC.M1.VV,
gr.VH1PP.F1.VV,
gr.VH1PP.M1.VV))

peakwidths <- width(combined.peaks)
combined.peaks <- combined.peaks[peakwidths  < 10000 & peakwidths > 20]
combined.peaks

md.DG_F1_VV <- read.table(
  file = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/metric_files/DG_F1_VV_per_barcode_metrics.csv",
  stringsAsFactors = FALSE,
  sep = ",",
  header = TRUE,
  row.names = 1
)

md.DG_F2_VV <- read.table(
  file = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/metric_files/DG_F2_VV_per_barcode_metrics.csv",
  stringsAsFactors = FALSE,
  sep = ",",
  header = TRUE,
  row.names = 1
)

md.DG_M1_VV <- read.table(
  file = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/metric_files/DG_M1_VV_per_barcode_metrics.csv",
  stringsAsFactors = FALSE,
  sep = ",",
  header = TRUE,
  row.names = 1
)

md.DG_M2_VV <- read.table(
  file = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/metric_files/DG_M2_VV_per_barcode_metrics.csv",
  stringsAsFactors = FALSE,
  sep = ",",
  header = TRUE,
  row.names = 1
)

md.NAC_F1_VV <- read.table(
  file = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/metric_files/NAC_F1_VV_per_barcode_metrics.csv",
  stringsAsFactors = FALSE,
  sep = ",",
  header = TRUE,
  row.names = 1
)

md.NAC_M1_VV <- read.table(
  file = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/metric_files/NAC_M1_VV_per_barcode_metrics.csv",
  stringsAsFactors = FALSE,
  sep = ",",
  header = TRUE,
  row.names = 1
)

md.VH1PP_F1_VV <- read.table(
  file = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/metric_files/VH1PP_F1_VV_per_barcode_metrics.csv",
  stringsAsFactors = FALSE,
  sep = ",",
  header = TRUE,
  row.names = 1
)

md.VH1PP_M1_VV <- read.table(
  file = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/metric_files/VH1PP_M1_VV_per_barcode_metrics.csv",
  stringsAsFactors = FALSE,
  sep = ",",
  header = TRUE,
  row.names = 1
)

md.DG_F1_VV <- md.DG_F1_VV[md.DG_F1_VV$atac_fragments > 500,]
md.DG_F2_VV <- md.DG_F2_VV[md.DG_F2_VV$atac_fragments > 500,]
md.DG_M1_VV <- md.DG_M1_VV[md.DG_M1_VV$atac_fragments > 500,]
md.DG_M2_VV <- md.DG_M2_VV[md.DG_M2_VV$atac_fragments > 500,]
md.NAC_F1_VV <- md.NAC_F1_VV[md.NAC_F1_VV$atac_fragments > 500,]
md.NAC_M1_VV <- md.NAC_M1_VV[md.NAC_M1_VV$atac_fragments > 500,]
md.VH1PP_F1_VV <- md.VH1PP_F1_VV[md.VH1PP_F1_VV$atac_fragments > 500,]
md.VH1PP_M1_VV <- md.VH1PP_M1_VV[md.VH1PP_M1_VV$atac_fragments > 500,]

frags.DG_F1_VV <- CreateFragmentObject(
  path = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/fragment_files/DG_F1_VV_atac_fragments.tsv.gz",
  cells = rownames(md.DG_F1_VV)
)

frags.DG_F2_VV <- CreateFragmentObject(
  path = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/fragment_files/DG_F2_VV_atac_fragments.tsv.gz",
  cells = rownames(md.DG_F2_VV)
)

frags.DG_M1_VV <- CreateFragmentObject(
  path = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/fragment_files/DG_M1_VV_atac_fragments.tsv.gz",
  cells = rownames(md.DG_M1_VV)
)

frags.DG_M2_VV <- CreateFragmentObject(
  path = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/fragment_files/DG_M2_VV_atac_fragments.tsv.gz",
  cells = rownames(md.DG_M2_VV)
)

frags.NAC_F1_VV <- CreateFragmentObject(
  path = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/fragment_files/NAC_F1_VV_atac_fragments.tsv.gz",
  cells = rownames(md.NAC_F1_VV)
)

frags.NAC_M1_VV <- CreateFragmentObject(
  path = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/fragment_files/NAC_M1_VV_atac_fragments.tsv.gz",
  cells = rownames(md.NAC_M1_VV)
)

frags.VH1PP_F1_VV <- CreateFragmentObject(
  path = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/fragment_files/VH1PP_F1_VV_atac_fragments.tsv.gz",
  cells = rownames(md.VH1PP_F1_VV)
)

frags.VH1PP_M1_VV <- CreateFragmentObject(
  path = "/data/johnsonko/SideProject/Raj/scRNA_Seq_Project_20221007/fragment_files/VH1PP_M1_VV_atac_fragments.tsv.gz",
  cells = rownames(md.VH1PP_M1_VV)
)

DG_F1_VV.counts <- FeatureMatrix(
  fragments = frags.DG_F1_VV,
  features = combined.peaks,
  cells = rownames(md.DG_F1_VV)
)

DG_F2_VV.counts <- FeatureMatrix(
  fragments = frags.DG_F2_VV,
  features = combined.peaks,
  cells = rownames(md.DG_F2_VV)
)

DG_M1_VV.counts <- FeatureMatrix(
  fragments = frags.DG_M1_VV,
  features = combined.peaks,
  cells = rownames(md.DG_M1_VV)
)

DG_M2_VV.counts <- FeatureMatrix(
  fragments = frags.DG_M2_VV,
  features = combined.peaks,
  cells = rownames(md.DG_M2_VV)
)

NAC_F1_VV.counts <- FeatureMatrix(
  fragments = frags.NAC_F1_VV,
  features = combined.peaks,
  cells = rownames(md.NAC_F1_VV)
)

NAC_M1_VV.counts <- FeatureMatrix(
  fragments = frags.NAC_M1_VV,
  features = combined.peaks,
  cells = rownames(md.NAC_M1_VV)
)

VH1PP_F1_VV.counts <- FeatureMatrix(
  fragments = frags.VH1PP_F1_VV,
  features = combined.peaks,
  cells = rownames(md.VH1PP_F1_VV)
)

VH1PP_M1_VV.counts <- FeatureMatrix(
  fragments = frags.VH1PP_M1_VV,
  features = combined.peaks,
  cells = rownames(md.VH1PP_M1_VV)
)

DG_F1_VV.chrom_assay <- CreateChromatinAssay(counts = DG_F1_VV.counts, fragments = frags.DG_F1_VV)
DG_F2_VV.chrom_assay <- CreateChromatinAssay(counts = DG_F2_VV.counts, fragments = frags.DG_F2_VV)
DG_M1_VV.chrom_assay <- CreateChromatinAssay(counts = DG_M1_VV.counts, fragments = frags.DG_M1_VV)
DG_M2_VV.chrom_assay <- CreateChromatinAssay(counts = DG_M2_VV.counts, fragments = frags.DG_M2_VV)
NAC_F1_VV.chrom_assay <- CreateChromatinAssay(counts = NAC_F1_VV.counts, fragments = frags.NAC_F1_VV)
NAC_M1_VV.chrom_assay <- CreateChromatinAssay(counts = NAC_M1_VV.counts, fragments = frags.NAC_M1_VV)
VH1PP_F1_VV.chrom_assay <- CreateChromatinAssay(counts = VH1PP_F1_VV.counts, fragments = frags.VH1PP_F1_VV)
VH1PP_M1_VV.chrom_assay <- CreateChromatinAssay(counts = VH1PP_M1_VV.counts, fragments = frags.VH1PP_M1_VV)

DG_F1_VV.atac <- CreateSeuratObject(DG_F1_VV.chrom_assay, assay = "ATAC", meta.data=md.DG_F1_VV)
DG_F2_VV.atac <- CreateSeuratObject(DG_F2_VV.chrom_assay, assay = "ATAC", meta.data=md.DG_F2_VV)
DG_M1_VV.atac <- CreateSeuratObject(DG_M1_VV.chrom_assay, assay = "ATAC", meta.data=md.DG_M1_VV)
DG_M2_VV.atac <- CreateSeuratObject(DG_M2_VV.chrom_assay, assay = "ATAC", meta.data=md.DG_M2_VV)
NAC_F1_VV.atac <- CreateSeuratObject(NAC_F1_VV.chrom_assay, assay = "ATAC", meta.data=md.NAC_F1_VV)
NAC_M1_VV.atac <- CreateSeuratObject(NAC_M1_VV.chrom_assay, assay = "ATAC", meta.data=md.NAC_M1_VV)
VH1PP_F1_VV.atac <- CreateSeuratObject(VH1PP_F1_VV.chrom_assay, assay = "ATAC", meta.data=md.VH1PP_F1_VV)
VH1PP_M1_VV.atac <- CreateSeuratObject(VH1PP_M1_VV.chrom_assay, assay = "ATAC", meta.data=md.VH1PP_M1_VV)

Idents(DG_F1_VV.atac) <- "DG_F1_VV"
Idents(DG_F2_VV.atac) <- "DG_F2_VV"
Idents(DG_M1_VV.atac) <- "DG_M1_VV"
Idents(DG_M2_VV.atac) <- "DG_M2_VV"
Idents(NAC_F1_VV.atac) <- "NAC_F1_VV"
Idents(NAC_M1_VV.atac) <- "NAC_M1_VV"
Idents(VH1PP_F1_VV.atac) <- "VH1PP_F1_VV"
Idents(VH1PP_M1_VV.atac) <- "VH1PP_M1_VV"

DG_F1_VV.atac$dataset <- "DG_F1_VV"
DG_F2_VV.atac$dataset <- "DG_F2_VV"
DG_M1_VV.atac$dataset <- "DG_M1_VV"
DG_M2_VV.atac$dataset <- "DG_M2_VV"
NAC_F1_VV.atac$dataset <- "NAC_F1_VV"
NAC_M1_VV.atac$dataset <- "NAC_M1_VV"
VH1PP_F1_VV.atac$dataset <- "VH1PP_F1_VV"
VH1PP_M1_VV.atac$dataset <- "VH1PP_M1_VV"

all.normalized.atac <- merge(DG_F1_VV.atac, y = c(DG_F2_VV.atac,DG_M1_VV.atac,DG_M2_VV.atac,NAC_F1_VV.atac,NAC_M1_VV.atac,VH1PP_F1_VV.atac,VH1PP_M1_VV.atac), add.cell.ids=c("DG_F1_VV","DG_F2_VV","DG_M1_VV","DG_M2_VV","NAC_F1_VV","NAC_M1_VV","VH1PP_F1_VV","VH1PP_M1_VV"), project="WT", merge.data=TRUE)

annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)
seqlevelsStyle(annotations) <- 'UCSC'
genome(annotations) <- "mm10"
Annotation(all.normalized.atac) <- annotations

all.normalized.atac <- RunTFIDF(all.normalized.atac)
all.normalized.atac <- FindTopFeatures(all.normalized.atac, min.cutoff = "q0")
all.normalized.atac <- RunSVD(all.normalized.atac)
all.normalized.atac <- RunUMAP(all.normalized.atac, reduction = "lsi", dims = 2:30, reduction.name = "umap.atac", reduction.key = "atacUMAP_")

**Comment by timoast** [comment]
For multiome data there's no need to perform ATAC/RNA integration and label transfer since the two measurements are made in the same cell. You should simply be able to merge all of your multiome datasets together, create the different dimension reductions for each assay (LSI on the ATAC assay and PCA on the RNA assay), and then run WNN on the merged object.
